File: MyAddViewEx/app/src/main/java/com/example/myaddviewex/MainActivity.java
package com.example.myaddviewex;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // 아이템 리스트를 준비
        // 변수는 하나만
        // 자료구조 - 하나의 변수에 관련된 데이터 타입을 통으로 관리하기 위해서 자료구조 or 배열을 사용한다.

        ArrayList<Car> carArrayList = new ArrayList<>();

        for (int i = 0; i < 100; i++) {
            carArrayList.add(new Car(i + "번째 자동차", i + "번째 엔진"));
        }

        LinearLayout container = findViewById(R.id.addViewContainer);

        // xml 파일을 들고 올려면 LayoutInflater
        LayoutInflater layoutInflater = LayoutInflater.from(this);

        for (int i = 0; i < carArrayList.size(); i++) {

            View itemView = layoutInflater.inflate(R.layout.item_view, null);
            TextView carTv = itemView.findViewById(R.id.carNameTextView);
            TextView engineTv = itemView.findViewById(R.id.engineTextView);
            carTv.setText(carArrayList.get(i).name);
            engineTv.setText(carArrayList.get(i).engine);

            container.addView(itemView);

        }


    }
}

File: Kotlin_Lab/Lab_20220428_example4/app/src/main/java/com/example/lab_20220428_example4/MainActivity.kt
package com.example.lab_20220428_example4

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import com.example.lab_20220428_example4.databinding.ActivityMainBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.consumeEach
import java.text.DecimalFormat

class MainActivity : AppCompatActivity() {

    var timeG = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        var binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        var timerView : TextView = findViewById(R.id.timerView)
        var startBtn : Button = findViewById(R.id.startBtn)
        var stopBtn : Button = findViewById(R.id.stopBtn)
        var pauseBtn : Button = findViewById(R.id.pauseBtn)
        var resumeBtn : Button = findViewById(R.id.resumeBtn)

        val channel = Channel<Int>()

        lateinit var backGroundScope: Job

        var mainScope = GlobalScope.launch(Dispatchers.Main) {
            channel.consumeEach {
                val df = DecimalFormat("00")
                var second = it / 60
                val minute = it / 3600
                val hour = it / 216000
                binding.timerView.text = "${df.format(hour)}:${df.format(minute)}:${df.format(second)}"
            }
        }

        startBtn.setOnClickListener {
            backGroundScope = CoroutineScope(Dispatchers.Default + Job()).launch {
                for (i in 1..2_000_000_000) {
                    timeG += 1
                    channel.send(timeG)
                    delay(10)
                }
            }
        }

        stopBtn.setOnClickListener {
            backGroundScope.cancel()
            timerView.text = "00:00:00"
            timeG = 0
        }

        pauseBtn.setOnClickListener {
            backGroundScope.cancel()
        }

        resumeBtn.setOnClickListener {
            val backGroundScope = CoroutineScope(Dispatchers.Default + Job())
            backGroundScope.launch {
                for (i in 1..2_000_000_000) {
                    timeG += 1
                    channel.send(timeG)
                    delay(10)
                }
            }
        }

    }
}
